feat: action for test containers

ramonpetgrave64 opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

https://github.com/sigstore/rekor-tiles/issues/38

Adds a new script for making it easier to launch a local fulcio and rekor, and tsa for testing within Github Actions.
It also creates a local truested_root.json and signing_config.json.

#### Testing

A new pre-submit called `test test_containers` runs the action, which has its own healthchecks for fulcio and rekor, and attempts to use the local containers for signing and verifying.

Also testing in https://github.com/sigstore/cosign/pull/4161

#### TODO

 - [x] Fix the `cosign verify-blob` step in the pre-submit https://github.com/sigstore/scaffolding/actions/runs/14392478215/job/40362163018?pr=1544#step:6:20
 - [x] Add rekor-v2
 - [x] shellcheck lints
 - [ ] Followup Issues/PRs about enabling local-development use-cases.
   - https://github.com/sigstore/scaffolding/issues/1555
 - [x] Add readme

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

jku commented:
Verify issue is fixed in https://github.com/sigstore/cosign/pull/4162 (can also be worked around by making sure rekor log index > 0)

jku commented:
Modifying my comment from previous PR:
* Let's consider what the provided API is here: in this PR the API is clearly the action... but that feels a bit incompatible with local uses case that I think is relevant: 
   * I would like to have a single command that starts the services and gives me a trusted root I can use -- I would love it if it also handles stopping services and allows configuration (e.g. `--enable-rekorv2`) but that can also be left for followup
   * I don't think we need the action at this point but am happy to hear arguments otherwise. Less API surface to maintain  means less work when we inevitably tweak and further improve this in the next weeks 
* Is there a reason for the external healthchecks in the action? Are the composed containers not reliably usable? 
* Is there a reason to not support fake oidc in the same way we support everything else (dockerfile + compose.yml ). I understand you would need to add those files into this PR but that seems preferable to handling fake oidc as a special case with different tools than everything else
* I believe we can remove the networking hacks from the script:  as far as I can tell we just don't need them, and was under the impression your solution (of using host `$(hostname)` ) with fakeoidc would work? If this is not the case, can we document the issues?

@ramonpetgrave64 let me know how you'd like to proceed: I'm available if you want to meet later today -- but I can also give refactoring this a try tomorrow if you have other things to do today.

ramonpetgrave64 commented:
@jku Right now, we're just getting something that we know will work well within Github Actions. We know we want the Action now for clients to test (cc @haydentherapper), so local use-cases can come later. I'll note that in the PR description.

ramonpetgrave64 commented on actions/test_containers/action.yml:
```diff
@@ -0,0 +1,70 @@
+      run: |
+        cat ${{ steps.run-containers.outputs.trusted-root }}
+        cat ${{ steps.run-containers.outputs.signing-config }}
+    - name: fulcio-healthcheck
+      shell: bash
+      run: curl -f http://localhost:5555/healthz
+    - name: rekor-healthcheck
+      shell: bash
+      run: curl -f http://localhost:3000/
+    - name: fakeoidc-healthcheck
+      shell: bash
+      run: curl -f ${{ steps.run-containers.outputs.oidc-url }}/.well-known/openid-configuration
+    - name: tsa-healthcheck
+      shell: bash
+      run: curl -f ${{ steps.run-containers.outputs.tsa-url }}/ping
```
Not needed, since run-containers.sh wil already have confirmed their healthy state. It's only nice to have.

ramonpetgrave64 replied:
removed

ramonpetgrave64 commented on actions/test_containers/action.yml:
```diff
@@ -0,0 +1,70 @@
+        go-version: "stable"
+        cache-dependency-path: "**/go.sum"
+    - uses: ko-build/setup-ko@d982fec422852203cfb2053a8ec6ad302280d04d # v0.8
+    # Pending https://github.com/sigstore/cosign-installer/issues/187.
+    # - uses: sigstore/cosign-installer@d7d6bc7722e3daa8354c50bcb52f4837da5e9b6a # v3.8.1
+    #   with:
+    #     cosign-release: main
```
the user should already have cosign installed, and we don't want to overwrite their installation.

jku replied:
> the user should already have cosign installed, and _we don't want to overwrite their installation_
 
 I tried to ask about this part: does using cosign installer  prevent the issue you mention?

ramonpetgrave64 replied:
Yes, but we can't use the cosign-installer yet because of the issue https://github.com/sigstore/cosign-installer/issues/187.

jku commented on .github/workflows/test-containers.yml:
```diff
@@ -0,0 +1,42 @@
+  pull_request:
+jobs:
+  test:
+    strategy:
+      matrix:
+        platform:
+          - ubuntu-latest
+    runs-on: ${{ matrix.platform }}
```
nit
```suggestion
    runs-on: ubuntu-latest
```

ramonpetgrave64 replied:
eventually we want to test on mac, which won't work yet. but ill remove this for now

jku commented:
> so local use-cases can come later.

Thank you for letting me know now. I disagree with this but let's just get something in even if it's unusable for local testing as it is.

jku reviewed (changes requested):
I'm repeating myself in the code comments, sorry about that. I just haven't seen you responding to the suggestions so I am making sure.

The action output is currently a bit excessive: multiple thousand lines. I think we can avoid some of the docker related spam in there

ramonpetgrave64 commented on actions/test_containers/run-containers.sh:
```diff
@@ -0,0 +1,116 @@
+done
+popd
+echo "building trusted root"
+if [[ "$USE_REKORV2_TRUSTED_ROOT" == "true" ]]; then
+  REKOR_ARGS=(--rekor-v2 http://localhost:3003 ~/rekor-tiles/tests/testdata/pki/ed25519-pub-key.pem)
+else
+  REKOR_ARGS=(--rekor-v1-url http://localhost:3000)
+fi
```
Yes, but the `cosign trusted-root create` doesn't seem like it can support rekor v1 and v2 trusted material simultaneously.

jku replied:
Can you be more specific? I don't see an issue?

ramonpetgrave64 replied:
Based on the `--help` output, I assumed that only one `--rekor-key` could be provided. But I see now that option is a string array. I'll change this.

ramonpetgrave64 commented on actions/test_containers/run-containers.sh:
```diff
@@ -0,0 +1,116 @@
+fi
+
+echo "setting up OIDC provider"
+pushd ./fakeoidc
+oidcimg=$(ko build main.go --local)
+docker network ls | grep fulcio_default || docker network create fulcio_default --label "com.docker.compose.network=fulcio_default"
+docker run -d --rm -p 8080:8080 --network fulcio_default --name fakeoidc "$oidcimg"
+oidc_ip=$(docker inspect fakeoidc | jq -r '.[0].NetworkSettings.Networks.fulcio_default.IPAddress')
```
Yes, it's not an easy to fit into the existing script. 
For the hostname, it doesn't seem like it's worth the extra effort.

ramonpetgrave64 replied:
using the compose file now

ramonpetgrave64 commented on actions/test_containers/action.yml:
```diff
@@ -0,0 +1,70 @@
+  oidc-url:
+    description: The url of the OIDC provider.
+    value: ${{ steps.run-containers.outputs.oidc-url }}
+  tsa-url:
+    description: The url of the Timestamp Authority (TSA).
+    value: ${{ steps.run-containers.outputs.tsa-url }}
```
Yes, it's needed for cosign's e2e tests

jku replied:
and why do cosigns e2e tests need timestamp url but don't need the other urls?

ramonpetgrave64 replied:
The cosign tests relied on using the TSA as a library. I'm changing them to use TSA as a service on the host.

 - https://github.com/sigstore/cosign/pull/4161/files#diff-c31efd77a0caecaeca4f91c9fc4a61f363fd07df32a48ac69a3fbc0cd5fcd5b4R185

jku replied:
This seems like cosign test problem -- why use the test setup based on trusted root when you are not willing to use the provided trusted root?

But like I said in another: I'm fine with leaving it in for now if it makes life easier in the other PR.  I just don't like that this is being designed as an extension of the cosign test suite (and not as a generic sigstore test system): if cosign tests are the main goal, then we could have left this in cosign git.

ramonpetgrave64 commented on .github/workflows/test-containers.yml:
```diff
@@ -0,0 +1,42 @@
+            --rekor-url http://localhost:3000 \
+            --certificate-identity foo@bar.com \
+            --certificate-oidc-issuer ${{ steps.run-test-containers.outputs.oidc-url }} \
+            --insecure-ignore-sct \
```
That was me trying to make the verify step work in github actions, before you discovered the bug. I'll remove this.

ramonpetgrave64 commented on actions/test_containers/run-containers.sh:
```diff
@@ -0,0 +1,116 @@
+       yq -i e '.networks={"default":{ "name":"fulcio_default","external":true }}' docker-compose.yml
+       yq -i e '.services.fulcio-server.networks=["default"]' docker-compose.yml
+    fi
+    ${docker_compose} up -d
+    echo -n "waiting up to 60 sec for system to start"
+    count=0
+    if [ "$repo" == "timestamp-authority" ] || [ "$repo" == "rekor-tiles" ]; then
+      target_healthy=1
+    else
+      target_healthy=3
+    fi
+    until [ "$(${docker_compose} ps | grep -c "(healthy)")" == "$target_healthy" ];
+    do
+        if [ $count -eq 18 ]; then
+           echo "! timeout reached"
+           exit 1
+        else
+           echo -n "."
+           sleep 10
+           (( count+=1 ))
+        fi
+    done
```
It's what we had before in the script copied from cosign, and `docker compose --wait` isn't available on some folks' older versions of docker.

jku replied:
I'm one of the people who grumbled and then upgraded compose: I think requiring the newer one is fine if we can make this script cleaner and easier to maintain.

ramonpetgrave64 replied:
changed.

ramonpetgrave64 commented on actions/test_containers/run-containers.sh:
```diff
@@ -0,0 +1,116 @@
+echo "starting services"
+export FULCIO_METRICS_PORT=2113
+export FULCIO_CONFIG=/tmp/fulcio-config.json
+for repo in rekor fulcio timestamp-authority rekor-tiles; do
```
That makes sense, but that would make the Action a lot more complicated. I want to avoid that complication.

ramonpetgrave64 replied:
Added to TODO

haydentherapper commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,111 @@
+
+echo "setting up kind"
+go install sigs.k8s.io/kind@v0.20.0
+kind create cluster
```
Why do we need `kind`?

ramonpetgrave64 replied:
I see it's actaully cosign-specific. removing.

 - https://github.com/sigstore/cosign/commit/8b0bcecfeb824e74fbad6607c719e77d628c3aab#diff-9e762d0df1502ee31a1de1a3bc67be6c2707ec3e6117be3147907a60ef5fb854R231

haydentherapper commented on actions/setup-sigstore-env/build-trusted-root.sh:
```diff
@@ -0,0 +1,117 @@
+
+[ -f trusted_root.json ] || [ -f signing_config.json ] && echo "trusted_root.json or signing_config.json already exist" && exit 1
+
+CMD="cosign trusted-root create"
```
Could we use the containerized version of Cosign so that we know Cosign is available and avoid prompting a user to install Cosign?

ramonpetgrave64 replied:
got it working.

haydentherapper commented on actions/setup-sigstore-env/action.yml:
```diff
@@ -0,0 +1,47 @@
+
+name: setup-sigstore-env
+outputs:
+  oidc-url:
```
Do we need oidc-url and tsa-url, since they're available in the SigningConfig?

ramonpetgrave64 replied:
They for the existing cosign e2e tests, they need to be available env variables.

jku replied:
This discussion is one of the reasons I was hoping to focus on the CLI first: the action API is clearly not quite "stable" yet:
* oidc-url is a bit of a special case as clients do need a way to get the token to sign with (and `curl $oidc/token` is a hacky but easy way to get it) and as the url is in signing config that is not well supported yet
* I think the long term solution is to provide the actual token (file) as an output... but we need some signingconfig support in clients before oidc-url can just be dropped 
* I would like to try to avoid exposing any urls directly -- this would make it clearer that trusted root / signing config is what clients should be looking at. But again, this hinges on some signingconfig support 
* Special cases like cosign TSA tests could always parse `trusted_root.json` (or hard code urls) if needed. 

So I suggest:
* let's leave `oidc-url` in there now but add `oidc-token` (filename that contains oidc token) as well and plan to drop oidc-url as signing config support gets better?
* we can leave tsa-url as well if that makes the cosign PR easier for now ... but users like cosign test suite should not expect to have tsa-url available in future: they should use trusted root and signing config

How does that sound?

jku replied:
to be clear, I think exposing all of the urls is an option too: it just makes it less obvious how this action is supposed to be used

ramonpetgrave64 replied:
added the URLs and file paths.

haydentherapper commented on .github/workflows/test-setup-sigstore-env.yml:
```diff
@@ -0,0 +1,36 @@
+name: test setup-sigstore-env
+on:
+  workflow_call:
```
Can we pair this down run when relevant files are updated?

ramonpetgrave64 replied:
fixed

jku commented on actions/setup-sigstore-env/action.yml:
```diff
@@ -0,0 +1,47 @@
+        go-version: "stable"
+        cache-dependency-path: "**/go.sum"
+    - uses: ko-build/setup-ko@d982fec422852203cfb2053a8ec6ad302280d04d # v0.8
+    # Pending https://github.com/sigstore/cosign-installer/issues/187.
+    # - uses: sigstore/cosign-installer@d7d6bc7722e3daa8354c50bcb52f4837da5e9b6a # v3.8.1
+    #   with:
+    #     cosign-release: main
+    - shell: bash
+      run: go install github.com/sigstore/cosign/v2/cmd/cosign@main
```
this cosign dependency is a little annoying for two reasons:
* it makes the local testing that bit more complicated (it's fun to run a 10 minute test process to find out I should have installed cosign first)
* this shadows any cosign install that the calling workflow already did and we probably don't want to break things like cosigns own tests like that 

some sort of container approach sounds better but maybe we leave that as followup?

ramonpetgrave64 replied:
changed to run the cosign container.

jku commented on actions/setup-sigstore-env/action.yml:
```diff
@@ -0,0 +1,47 @@
+runs:
+  using: composite
+  steps:
+    - uses: actions/setup-go@f111f3307d8850f501ac008e886eec1fd1932a34 # v5.3.0
+      with:
+        go-version: "stable"
+        cache-dependency-path: "**/go.sum"
+    - uses: ko-build/setup-ko@d982fec422852203cfb2053a8ec6ad302280d04d # v0.8
```
this dependency on ko seems like it would go away if used compose for fakeoidc?

ramonpetgrave64 replied:
changed to use compose

jku commented:
Summarizing my requests/questions:
* action output is still excessive
* I still think fakeoidc should be composed if possible: this would remove `ko` and `yq` from the dependencies and should simplify the script and make using it locally at least possible: doing that right now is quite painful.
* requiring a fairly recent docker compose seems reasonable: this means we can stop doing any manual health checks
* cosign install has some corner cases (but we could leave improvements to followups)
* outputs: I think adding `oidc-token` makes sense. We can leave the current url outputs in there for now (especially oidc-url is tricky to remove yet) but I suggest we plan to remove them for clarity: Ideally clients operate using trusted root and signing config (see the specific thread for details)

jku commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,107 @@
+pushd ./fakeoidc
+docker build . -t fakeoidc
+docker network ls | grep fulcio_default || docker network create fulcio_default --label "com.docker.compose.network=fulcio_default"
+docker network ps | grep fakeoidc || docker run -d --rm -p 8080:8080 --network fulcio_default --name fakeoidc fakeoidc
```
I've already suggested replacing this but just for completeness `docker network ps` is not a thing and will always fail

ramonpetgrave64 replied:
fixed

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,107 @@
+EOF
+popd
+
+pushd "$HOME"
```
changed to use temp folder

ramonpetgrave64 commented on .github/workflows/test-containers.yml:
```diff
@@ -0,0 +1,42 @@
+        with:
+          persist-credentials: false
+      - id: run-test-containers
+        uses: ./actions/test_containers
```
renamed.

ramonpetgrave64 commented:
@jku Added lots of changes addressing your comments. With the new changes, there's just one cosign e2e test failing  that you may be able to fix, while I also try to fix it.

 - https://github.com/sigstore/cosign/actions/runs/14541725399/job/40800903020?pr=4161#step:9:457

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -ex
```
If the user's goal is only to run the containers, then sourcing isn't necessary. Changing it to invoke `bash -ex ./run-containers.sh` from the action, instead.

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,128 @@
+export TSA_URL="http://${HOST}:3004"
+export CT_LOG_KEY="$WORKDIR/fulcio/config/ctfe/pubkey.pem"
+GITHUB_ACTIONS="${GITHUB_ACTIONS:-false}"
+if [[ "$GITHUB_ACTIONS" != "false" ]]; then
```
These things need to be in the script because the variables we need to export wouldn't be available outside the scrtipt.

jku replied:
OK. This is usually the reason GH Actions shell scripts are made sourcable but I suppose I assumed and that was not the case here

jku replied:
yeah exactly that was what I expected was the idea when I saw the script was made sourcable, but I'm not requesting it be done in this PR

ramonpetgrave64 replied:
Now exporting from the action.yml, instead of the scripts.

ramonpetgrave64 commented on actions/setup-sigstore-env/fakeoidc/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY . .
+
+RUN go mod download
+# RUN go vet -v
+RUN go test -v
```
removed.

ramonpetgrave64 commented on actions/setup-sigstore-env/action.yml:
```diff
@@ -0,0 +1,49 @@
+  signing-config:
+    description: Path to the singning config json file.
+    value: ${{ steps.run-containers.outputs.signing-config }}
+  clone-dir:
+    description: Path to the directory where the service repositories are cloned.
+    value: ${{ steps.run-containers.outputs.clone-dir }}
```
cosign needs this for some e2e tests that need the fulcio repo's embedded keys, and other users of the action may have use for that.

ramonpetgrave64 replied:
embedded.

ramonpetgrave64 commented on actions/setup-sigstore-env/action.yml:
```diff
@@ -0,0 +1,49 @@
+
+name: setup-sigstore-env
+outputs:
+  ct-log-key:
+    description: Path to the CT log public key.
+    value: ${{ steps.run-containers.outputs.ct-log-key }}
```
Which outputs should we remove?

jku replied:
This is one of the things that would have been nice to have in a comment: _why did you have to add two new outputs that should not be needed, that were not needed in the previous iteration?_

I think the answer is that because you are now using a temporary directory, you can't hard code the locations in the cosign tests anymore? So if we later make this not use a temporary dir  (or make the temporary dir name a constant one), the outputs are not needed?

jku replied:
I've let a comment in the cosign PR: embedding the fulcio/config/ctfepubkey.pem test key in the source there seems fine  to me.

jku commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -15,7 +15,7 @@
 
 echo "setting up OIDC provider"
-pushd ./fakeoidc
+pushd ./fakeoidc || exit
```
`exit` has the same issue as "set -e" in a script that is sourced...

ramonpetgrave64 replied:
ok. `|| return` also works and won't exit the shell.

ramonpetgrave64 commented on actions/setup-sigstore-env/build-trusted-root.sh:
```diff
@@ -0,0 +1,127 @@
+echo "Wrote trusted_root.json & signing_config.json"
+
+CWD="$(pwd)"
+export TRUSTED_ROOT="$CWD/trusted_root.json"
+export SIGNING_CONFIG="$CWD/signing_config.json"
+GITHUB_ACTIONS="${GITHUB_ACTIONS:-}"
+if [[ -n "$GITHUB_ACTIONS" ]]; then
+  # GitHub action env and outputs
+  echo "TRUSTED_ROOT=$TRUSTED_ROOT" >> "$GITHUB_ENV"
+  echo "trusted-root=$TRUSTED_ROOT" >> "$GITHUB_OUTPUT"
+
+  echo "SIGNING_CONFIG=$SIGNING_CONFIG" >> "$GITHUB_ENV"
+  echo "signing-config=$SIGNING_CONFIG" >> "$GITHUB_OUTPUT"
+fi
```
resolved https://github.com/sigstore/scaffolding/pull/1544#discussion_r2054262566

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,128 @@
+EOF
+popd
+
+WORKDIR=$(mktemp -d)
```
I'll keep it this way for now by default.

ramonpetgrave64 commented on actions/setup-sigstore-env/build-trusted-root.sh:
```diff
@@ -0,0 +1,127 @@
+done
+
+# cosign needs both +r and +x permissions
+chmod a+rx -R "$WORKDIR"
```
added

ramonpetgrave64 commented on .github/workflows/test-setup-sigstore-env.yml:
```diff
@@ -0,0 +1,47 @@
+            --rekor-url http://localhost:3000 \
+            --fulcio-url http://localhost:5555 \
+            --oidc-issuer ${{ steps.setup-sigstore-env.outputs.oidc-url }} \
+            --identity-token=$(curl -f ${{ steps.setup-sigstore-env.outputs.oidc-url }}/token) \
```
changed the action to output the token content.

haydentherapper replied:
This would occur if the identity token is missing or is invalid - https://github.com/sigstore/cosign/blob/d6c40c940fb04cd62156a69b32f46df5a1012bfa/cmd/cosign/cli/fulcio/fulcio.go#L149-L152

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,128 @@
+      # then quickly attach the fakeoidc container to the fulcio_default network.
+      docker network inspect fulcio_default | grep fakeoidc || docker network connect --alias "$HOST" fulcio_default fakeoidc
+    fi
+    # sometimes the services only become healthy after first becoming unhealthy, so we run this command twice.
+    docker compose up --wait || docker compose up --wait
```
fixing in https://github.com/sigstore/rekor/pull/2459

ramonpetgrave64 replied:
fixed in https://github.com/sigstore/rekor/pull/2459

haydentherapper commented on .github/workflows/test-setup-sigstore-env.yml:
```diff
@@ -0,0 +1,47 @@
+          persist-credentials: false
+      - id: setup-sigstore-env
+        uses: ./actions/setup-sigstore-env
+      # Pending https://github.com/sigstore/cosign-installer/issues/187.
```
I cut a new release https://github.com/sigstore/cosign-installer/releases/tag/v3.8.2

It would be nice if we could reuse the Cosign container downloaded in the script rather than download the binary, we can follow up on this later.

haydentherapper reviewed (commented):
Just some small comments!

Can you create an issue tracking follow ups from this PR? Any cleanup in Cosign testing, removing GHA references from the script, testing on other OSs, anything else?

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,124 @@
+# limitations under the License.
+
+echo "setting up OIDC provider"
+pushd ./fakeoidc || return
```
it's so the script can exit early upon error, without using `set -e`. This allows us to "source" the script without our terminals quitting upon error.

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,124 @@
+    if [[ "$repo" == "fulcio" ]]; then
+      # create the fulcio_default network by running `compose up`.
+      docker compose up -d
+      # then quickly attach the fakeoidc container to the fulcio_default network.
```
Yes, even if the attachment happens after the container is fully launched, it's still okay, only that there is an initial error in the logs of fulcio_server.

ramonpetgrave64 commented on actions/setup-sigstore-env/run-containers.sh:
```diff
@@ -0,0 +1,124 @@
+echo "setting up OIDC provider"
+pushd ./fakeoidc || return
+docker compose up --wait
+# the faeoidc container's hostname must be the same, both from within fulcio and from this host machine.
```
fixed

ramonpetgrave64 commented on actions/setup-sigstore-env/fakeoidc/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+RUN CGO_ENABLED=0 go build -o /go/bin/app
+
+FROM gcr.io/distroless/base-debian12:debug@sha256:3a59a8d10471fc8487fd2ca93746b0195ed4c3236c14fe8412cf7b2ec4b8c1f3
```
added

haydentherapper commented on actions/setup-sigstore-env/build-trusted-root.sh:
```diff
@@ -0,0 +1,122 @@
+
+WORKDIR=$(mktemp -d)
+# run cosign as a container with the current user permissions. This script will copy files into $WORKDIR.
+COSIGN_CMD="docker run --user=$(id -u):$(id -g) --rm -v $WORKDIR/:$WORKDIR/ ghcr.io/sigstore/cosign/cosign@sha256:e82eb6d42ccb6bc048d8d9e5e598e4d5178e1af6c00e54e02c9b0569c5f3ec11"
```
Hm, not sure why this is the case, it might be that `cosign copy` isn't properly copying multiplatform images. `gcr.io/projectsigstore/cosign:latest` has a linux/amd64 build.

ramonpetgrave64 replied:
I made a new Dockerfile.cosign with the full digest. dependabot should be auto-updating that.

jku reviewed (approved):
Nice improvements on the runtime: 
* 13 secs on my laptop when containers are running already 
* when starting cold (as in containers are not running but have been built before) it's 120 secs (50 secs of that is rekor v1 startup)

 I think I'm ok with merging this -- it  still has some issues but I think iterating on this PR is not the best way forward. let's merge and improve in other prs.

My general feedback remains same: _please document decisions and design changes in PR comments_. Seeing that review comments have been marked resolved is not the same as seeing an explanation of how you decided to proceed and why that decision was made.

I also wanted to mention that the review comments on the shell script sourcing have been technical nitpicks because there was never a design decision on the table: No-one tried to answer the question _should the script be sourcable or not_  because the decision was never brought up: we just noticed the README saying we should source the script...

The remaining issue with the "sourced shell script" design is that the error handling is now missing for most commands -- this is a common issue shared by all sourced script files. I think we should consider this a bug in the script but I am fine handling that as a separate issue.

jku commented:
> The remaining issue with the "sourced shell script" design is that the error handling is now missing for most commands -- this is a common issue shared by all sourced script files. I think we should consider this a bug in the script but I am fine handling that as a separate issue.

ramon correctly pointed out that the action runs the source in a shell that has "set -e" -- so it should fail visibly. That makes this much less of a problem
